diff --git a/lib/handlers/element.js b/lib/handlers/element.js
index 9bf3e0436a9b4d945e92e8914bac181cd83583da..aea581eae2ae5cb44619731d225b82024c0b446c 100644
--- a/lib/handlers/element.js
+++ b/lib/handlers/element.js
@@ -105,7 +105,7 @@ export function element(node, state) {
               method: false,
               shorthand: false,
               computed: false,
-              key: {type: 'Identifier', name: cssProp},
+              key: cssProp.includes('-') ? {type: 'Literal', value: cssProp} : {type: 'Identifier', name: cssProp},
               value: {type: 'Literal', value: String(styleObject[cssProp])},
               kind: 'init'
             })