diff --git a/lib/handlers/element.js b/lib/handlers/element.js
index f0718a28c971a1339a8a94bd38b80b91c3de4c6e..d2355e71e8a6d09c351795e93db4a4a48661ea14 100644
--- a/lib/handlers/element.js
+++ b/lib/handlers/element.js
@@ -65,7 +65,7 @@ export function element(node, state) {
         continue
       }
 
-      prop = info.space
+      prop = info.space && prop !== 'className'
         ? hastToReact[info.property] || info.property
         : info.attribute
 
@@ -101,7 +101,10 @@ export function element(node, state) {
           }
         }
 
-        attributeValue = {
+        attributeValue = typeof value === 'string' ? {
+          type: "Literal",
+          value,
+        } : {
           type: 'JSXExpressionContainer',
           expression: {type: 'ObjectExpression', properties: cssProperties}
         }